File: sdks/canary/typescript/scenarios/d-change-password.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// D-CHANGE-PASSWORD canary — TypeScript SDK

import { LLMSafeSpaces, AuthError } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, rawDo } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  const rotateEmail = process.env.LLMSAFESPACES_ROTATE_EMAIL || 'canary-rotate@llmsafespaces.test';
  const rotatePassword = process.env.LLMSAFESPACES_ROTATE_PASSWORD || 'canary-rotate-password!';
  const newPassword = 'canary-rotate-newpw!';

  const c = new LLMSafeSpaces({
    baseUrl: cfg.apiUrl,
    credentials: { email: rotateEmail, password: rotatePassword },
    timeout: 60000,
    fetch: nodeFetch as any,
  });

  const [ok, secret] = await r.assertNoError(
    () => c.secrets.create({ name: 'canary-pw-change-test', type: 'env-secret', value: 'pw-change-secret', metadata: { var_name: 'CANARY_TS_VAR' } }),
    'create-secret: no error');
  if (!ok || !secret) return;

  try {
    await r.assertNoError(
      () => c.account.changePassword(rotatePassword, newPassword),
      'change-password: no error');

    const loginBody = Buffer.from(JSON.stringify({ email: rotateEmail, password: newPassword }));
    const [s1, b1] = await rawDo('POST', `${cfg.apiUrl}/api/v1/auth/login`, '', loginBody);
    r.assert(s1 === 200, 'login-new-password: 200', `got ${s1}`);
    r.assert(hasField(b1, 'token'), 'login-new-password: has token');

    const loginBodyOld = Buffer.from(JSON.stringify({ email: rotateEmail, password: rotatePassword }));
    const [s2] = await rawDo('POST', `${cfg.apiUrl}/api/v1/auth/login`, '', loginBodyOld);
    r.assert(s2 === 401, 'login-old-password: 401', `got ${s2}`);

    const c2 = new LLMSafeSpaces({
      baseUrl: cfg.apiUrl,
      credentials: { email: rotateEmail, password: newPassword },
      timeout: 60000,
      fetch: nodeFetch as any,
    });
    const [ok3, revealed] = await r.assertNoError(
      () => c2.secrets.reveal(secret.id, newPassword),
      'reveal-with-new-password: no error');
    if (ok3 && revealed) {
      r.assert(revealed.value === 'pw-change-secret', 'reveal-with-new-password: value correct',
        revealed.value);
    }

    await r.assertNoError(
      () => c2.account.changePassword(newPassword, rotatePassword),
      'change-password-back: no error');

    await r.assertError(
      () => c.account.changePassword('wrong-password', newPassword),
      'change-wrong-old-password: error');

    const [ok4] = await r.assertNoError(
      () => c.account.changePassword(rotatePassword, 'short').catch((e: any) => {
        if (e?.status === 400 || e instanceof AuthError) throw e;
        throw e;
      }),
      'change-too-short-password: error');
    if (ok4 === false) r.ok('change-too-short-password: caught error');

  } finally {
    try {
      const cCleanup = new LLMSafeSpaces({
        baseUrl: cfg.apiUrl,
        credentials: { email: rotateEmail, password: rotatePassword },
        timeout: 30000,
        fetch: nodeFetch as any,
      });
      await cCleanup.secrets.delete(secret.id);
    } catch {}
  }
}

function hasField(body: Buffer, field: string): boolean {
  try {
    const obj = JSON.parse(body.toString());
    return field in obj;
  } catch { return false; }
}

async function main() {
  const r = new Runner('change-password');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });

File: sdks/canary/python/scenarios/d_change_password.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""D-CHANGE-PASSWORD canary — Python SDK"""

from __future__ import annotations

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import Runner, Config, config_from_env
from llmsafespaces import LLMSafeSpaces, AuthError


def run(r: Runner, cfg: Config) -> None:
    email = os.environ.get("LLMSAFESPACES_ROTATE_EMAIL", "canary-rotate@llmsafespaces.test")
    password = os.environ.get("LLMSAFESPACES_ROTATE_PASSWORD", "canary-rotate-password!")
    new_password = "canary-new-pw-12345!"

    c = LLMSafeSpaces(cfg.api_url, email=email, password=password, timeout=30.0)
    secret_id = None
    try:
        ok_me, me = r.assert_no_error(lambda: c.auth.me(), "login: no error")
        if not ok_me:
            return

        ok, secret = r.assert_no_error(
            lambda: c.secrets.create(
                name="canary-py-chpw", type="env-secret", value="chpw-test-value", metadata={"var_name": "CANARY_PY_VAR"}
            ),
            "create-secret: no error",
        )
        if not ok:
            return
        secret_id = secret.id

        r.assert_no_error(
            lambda: c.account.change_password(password, new_password),
            "change-password: no error",
        )

        c2 = LLMSafeSpaces(cfg.api_url, email=email, password=new_password, timeout=30.0)
        ok2, me2 = r.assert_no_error(lambda: c2.auth.me(), "login-new-pw: no error")
        r.assert_(ok2, "login-new-pw: succeeds")

        r.assert_error(
            lambda: LLMSafeSpaces(
                cfg.api_url, email=email, password=password, timeout=10.0
            ).auth.me(),
            "login-old-pw: error (401)",
        )

        ok3, val = r.assert_no_error(
            lambda: c2.secrets.reveal(secret_id, new_password),
            "reveal-new-pw: no error",
        )
        if ok3:
            r.assert_(val == "chpw-test-value", "reveal-new-pw: value matches", repr(val))

        r.assert_no_error(
            lambda: c2.account.change_password(new_password, password),
            "change-back: no error",
        )

        # Negative: wrong old password
        r.assert_error(
            lambda: c.account.change_password("definitely-wrong-xyz", new_password),
            "change-wrong-old: error",
        )

        # Negative: short new password
        r.assert_error(
            lambda: c.account.change_password(password, "short"),
            "change-short-new: error",
        )

    finally:
        if secret_id:
            try:
                c.secrets.delete(secret_id)
            except Exception:
                try:
                    c2.secrets.delete(secret_id)
                except Exception:
                    pass


if __name__ == "__main__":
    r = Runner("change-password")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())
